src/vl-tooltip.js

```diff
@@ -0,0 +1,624 @@
+    vgView.on("mousemove", function(event, item) {
+      update(event, item, options);
+    });
+
+    // clear tooltip on mouse out
+    vgView.on("mouseout", function(event, item) {
+      clear();
+    });
+  };
+
+  /* Mapping from fieldDef.type to formatType */
+  var formatTypeMap = {
+    "quantitative": "number",
+    "temporal": "time",
+    "ordinal": undefined,
+    "nominal": undefined
+  }
+
+  /**
+  * (Vega-Lite only) Supplement options with vlSpec
+  * if options.showAllFields is true or undefined, vlSpec will supplement
+  * options.fields with all fields in the spec
+  * if options.showAllFields is false, vlSpec will only supplement existing fields
+  * in options.fields
+  */
+  function supplementOptions(options, vlSpec) {
+    // fields to be supplemented by vlSpec
+    var supplementedFields = [];
+
+    var timeFormat = vlSpec.config ? vlSpec.config.timeFormat : undefined;
+    var numberFormat = vlSpec.config ? vlSpec.config.numberFormat : undefined;
+
+    // if showAllFields is true or undefined, supplement all fields in vlSpec
+    if (options.showAllFields !== false) {
+      vl.spec.fieldDefs(vlSpec).forEach(function(fieldDef){
+        // get a field in options that matches the fieldDef
+        var field = getField(options.fields, fieldDef);
+
+        // supplement the field with fieldDef, timeFormat and numberFormat
+        var supplementedField = supplementField(field, fieldDef, timeFormat, numberFormat);
+
+        supplementedFields.push(supplementedField);
+      });
+    }
+    // if showAllFields is false, only supplement existing fields in options.fields
+    else {
+      if (options.fields && options.fields.length > 0) {
+        options.fields.forEach(function(field) {
+          // get the fieldDef in vlSpec that matches the field
+          var fieldDef = getFieldDef(vl.spec.fieldDefs(vlSpec), field);
+
+          // supplement the field with fieldDef, timeFormat and numberFormat
+          var supplementedField = supplementField(field, fieldDef, timeFormat, numberFormat);
+
+          supplementedFields.push(supplementedField);
+        })
+      }
+    }
+
+    // TODO(zening): supplement binned fields
+
+    options.fields = supplementedFields;
+
+    return options;
+  }
+
+  /**
+  * Find a field in fields that matches a fieldDef
+  * If the fieldDef is aggregated, find a field that matches the field name and
+  * the aggregation of the fieldDef.
+  * If the fieldDef is not aggregated, find a field that matches the field name.
+  * @return the matching field, or undefined if no match was found
+  */
+  function getField(fields, fieldDef) {
+    if (!fieldDef || !fields || fields.length <= 0) return;
+
+    var matchedField = undefined;
+
+    // if aggregate, match field name and aggregate operation
+    if (fieldDef.aggregate) {
+      // try find the perfect match: field name equals, aggregate operation equals
+      fields.forEach(function(field) {
```

Comment:
Do you always need to run through every field?  If not, use normal `for`-loop and `break`.